python/tests/integration/server_manager.py

```diff
@@ -103,29 +96,99 @@ def handle_call(self, context, connection):
             if exp.matches(context.message):
                 exp.respond(context, connection)
 
+    def __enter__(self):
+        self.start()
+        return self
+
+    def __exit__(self, *args):
+        self.stop()
+
+    def start(self):
+        assert self.thread is None, 'server already started'
+        self.thread = threading.Thread(target=self.serve)
```

Comment:
Since this class is not meant to be used directly and expects child classes to implement `serve`, can you add a dummy `serve` method to this class that `raise`s `NotImplementedError`?